File: python/src/persian_text_guard/_utf16.py
r"""The UTF-16 view the matcher runs on, and the mapping of its positions back to code points.

.NET reads text as UTF-16 code units, and decides tokens by testing single units: half of a surrogate
pair is not a letter. Python strings are code points, so a port that tested them directly would read
U+20000 (a CJK letter) as one letter where .NET sees two surrogate units, and would find different
words (spec 004, research R2). For example, .NET flags ``kir𠀀`` as ``kir`` at UTF-16 index 0,
length 5, and censors it to ``****``: the two surrogates are not letters, so they join the word.

So every public entry point converts the caller's text to a *view*: a ``str`` in which each code
point above U+FFFF is replaced by its two surrogate code points. Inside the package, ``text[i]`` is
then UTF-16 unit *i*, exactly as ``charCodeAt(i)`` is in the JavaScript port, and ``len(text)`` is the
unit count. The view never leaves the package:

- match positions are converted back to code points of the caller's string with :class:`PositionMap`
  (``kir𠀀`` gives index 0, length 4);
- ``censor`` splices the caller's own string, so text outside a match is returned exactly as given,
  including surrogates the caller split themselves;
- normalizer and tokenizer output is converted with :func:`from_view`, which recombines every valid
  surrogate pair. There the caller's own split pair ``"\ud840\udc00"`` becomes the one code point
  U+20000, which is what .NET does when its UTF-16 result is decoded.
"""

from __future__ import annotations

import re

from ._unicode import recombine, split_supplementary

SUPPLEMENTARY = re.compile("[\U00010000-\U0010ffff]")


def to_view(text: str) -> str:
    """The UTF-16 view of ``text``: ``text`` itself unless it has a code point above U+FFFF."""
    if text.isascii() or SUPPLEMENTARY.search(text) is None:
        return text
    return split_supplementary(text)


def from_view(view: str) -> str:
    """A view turned back into code points: every valid surrogate pair becomes one code point."""
    return recombine(view)


class PositionMap:
    """Converts positions in a view to positions in the caller's string.

    It is the identity when the view is the caller's string. Otherwise it holds, for each unit index of
    the view, the code point index of the character it belongs to; the second unit of a pair maps to
    its pair's start.
    """

    __slots__ = ("_starts",)

    _starts: list[int] | None

    def __init__(self, text: str, view: str) -> None:
        """Build the map for ``text`` and its ``view``."""
        if view is text or len(view) == len(text):
            self._starts = None
            return

        # Each code point contributes one entry per unit, each equal to its own index.
        starts: list[int] = []
        for index, c in enumerate(text):
            starts.append(index)
            if c > "\U0000ffff":
                starts.append(index)
        self._starts = starts

    def to_code_points(self, unit_index: int, unit_length: int) -> tuple[int, int]:
        """The ``(index, length)`` in code points of the view region ``(unit_index, unit_length)``."""
        starts = self._starts
        if starts is None:
            return unit_index, unit_length
        start = starts[unit_index]
        end = starts[unit_index + unit_length - 1] + 1
        return start, end - start

File: rust/src/utf16.rs
//! The UTF-16 view the matcher runs on, and the conversion of its positions to bytes (research R2).
//!
//! .NET reads text as UTF-16 units and decides tokens by testing single units: half of a surrogate pair
//! is not a letter. A Rust `&str` is UTF-8, so a port testing `char`s would read U+20000 (a CJK letter)
//! as one letter where .NET sees two surrogate units, and would find different words. So the matcher,
//! ported line by line from TypeScript, runs on a *view*: the message's UTF-16 units, where `text[i]` is
//! unit *i* as `charCodeAt(i)` is in TypeScript.
//!
//! The view never leaves the crate. Match regions are converted to byte offsets of the caller's string
//! with [`ByteMap`]; `censor` splices the caller's own string; normalizer output is decoded with
//! [`from_view`]. For example, .NET flags `kir𠀀` as `kir` at units 0..5 (the two surrogates are not
//! letters, so they join the word); here that region is bytes 0..7 of the string, all of it.

/// The UTF-16 view of `text`.
pub(crate) fn to_view(text: &str) -> Vec<u16> {
    text.encode_utf16().collect()
}

/// A view decoded back to UTF-8. No lone surrogate is produced inside the crate, but a lossy decode can
/// never panic either way.
pub(crate) fn from_view(view: &[u16]) -> String {
    String::from_utf16_lossy(view)
}

/// Converts regions of a message's view, in UTF-16 units, to byte regions of the message.
///
/// The identity for ASCII messages. Otherwise it holds, for each unit, the byte offset of the character
/// it belongs to. Regions never split a surrogate pair, so every converted offset is a character
/// boundary.
pub(crate) struct ByteMap<'a> {
    text: &'a str,
    starts: Option<Vec<usize>>,
}

impl<'a> ByteMap<'a> {
    /// The map for `text`.
    pub(crate) fn new(text: &'a str) -> Self {
        if text.is_ascii() {
            return Self { text, starts: None };
        }

        let mut starts = Vec::with_capacity(text.len());
        for (offset, c) in text.char_indices() {
            starts.push(offset);
            if c.len_utf16() == 2 {
                starts.push(offset);
            }
        }

        Self {
            text,
            starts: Some(starts),
        }
    }

    /// The byte region `(start, len)` of the unit region `(index, length)`. `length` is at least 1.
    pub(crate) fn to_bytes(&self, index: usize, length: usize) -> (usize, usize) {
        let Some(starts) = &self.starts else {
            return (index, length);
        };

        let start = starts.get(index).copied().unwrap_or(self.text.len());
        let last = starts
            .get(index + length.max(1) - 1)
            .copied()
            .unwrap_or(self.text.len());
        let end = last + self.text[last..].chars().next().map_or(0, char::len_utf8);
        (start, end.max(start) - start)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn a_supplementary_letter_is_two_units_and_four_bytes() {
        let text = "kir\u{20000}";
        assert_eq!(to_view(text).len(), 5);
        let map = ByteMap::new(text);
        assert_eq!(map.to_bytes(0, 5), (0, 7));
        assert_eq!(map.to_bytes(3, 2), (3, 4));
        assert_eq!(map.to_bytes(0, 3), (0, 3));
    }

    #[test]
    fn ascii_is_the_identity() {
        let map = ByteMap::new("hello kir");
        assert_eq!(map.to_bytes(6, 3), (6, 3));
    }

    #[test]
    fn persian_letters_are_two_bytes() {
        let text = "\u{1F600} کیر";
        let map = ByteMap::new(text);
        assert_eq!(map.to_bytes(3, 3), (5, 6));
        assert_eq!(&text[5..11], "کیر");
    }

    #[test]
    fn views_round_trip() {
        let text = "سلام \u{1F600} kir";
        assert_eq!(from_view(&to_view(text)), text);
    }
}
